### PR TITLE
Implement bipolar scales

### LINKED ISSUE #1148: Two-ended scales
Two-ended scales should be added to enable questions like "How do you rate the difficulty?" on a scale from "too low" via "just right" to "too high".

We need to discuss:
- How should the voting buttons and the results distribution look like? Both ends of the scale are negative, the middle option is positive - this is in contrast to all other rating question we currently have.
- How can the scales be customized, i.e., that they can be used for questions asking for "low/high" as well as for "less/more", etc.

### PR DESCRIPTION
**Pull request by pixunil**
Resolves #1148.

### Status
All parts are ready for review. Screenshots needs replacements.

### Test data
I added or changed several questions. The answers were generated by taking the course distribution and randomly selecting from this population. For information, [see the full code](https://gist.github.com/pixunil/fcdb116dba060e8dc4956c103a87f4c4).

### Redesign
There were five buttons for likert scales. But with bipolar options, there are a total of seven options plus one "no answer", which took significantly more space. Thus, the vote buttons got more space (and the "no answer" shrunk) and now align with the left border of the text answers. Also, the `<table>` was too inflexible so it was replaced by `.col`. Also, all screen sizes from `md` are now properly displayed. For `xl` and `sm`, a wrapping of voting buttons is inevitable. To slightly improve this situation, a small margin keeps the buttons vertically separated.

### Screenshots
likert-choices
bipolar-choices
yes-no-choices
bipolar-results

### CODE REVIEW

**Comment by He3lixxx** [comment]
Whoa, test_data.json +30k lines? How many answers did you add there?

File: evap/evaluation/models.py [hunk 1]
```diff
@@ -718,39 +746,161 @@ def answer_class(self):
...
+        _("5"),
+        _("no answer")
+    ),
+    Question.EASY_DIFFICULT: BipolarChoices(
+        _("Easy"), _("Difficult"),
+        _("Way too\neasy"),
+        _("Too easy"),
+        _("Slightly too\neasy"),
+        _("Ideal"),
+        _("Slightly too\ndifficult"),
```
[the remark this answers is not in the document]
Comment (janno42) [reply evap/evaluation/models.py]: actually, the line should also break at shorter texts like `Too\neasy` to have a consistent look on the buttons (top always showing the "amount", bottom the two poles we're dealing with)

File: evap/static/css/_vote-colors.scss [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+        color: $white;
+        @include bar-shadow($medium-gray);
+    }
+    &:not(:hover) .adjective, &:hover .fas {
+            display: none;
+        }
+    }
+
+.distribution-bar {
```
Comment (janno42) [inline evap/static/css/_vote-colors.scss]: the file name is not really explaining what it contains, i wouldn't have expected this for example and it has a lot more than just colors. so what about `vote-elements` or similar?

File: evap/static/css/_vote-colors.scss [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+}
+.vote-btn-text {
+    position: relative;
+    top: -1px;
+}
+
+@each $name, $color in $vote-colors {
+    .vote-btn-#{$name} {
+        background: linear-gradient(to top, 5% $color, 5% transparent);
+        background-color: change-color($color, $lightness: 96%);
```
Comment (janno42) [inline evap/static/css/_vote-colors.scss]: maybe `97%`?

File: evap/static/css/_vote-colors.scss [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        color: darken($color, 25%);
+
+        &:hover {
+            background-color: change-color($color, $lightness: 80%);
```
Comment (janno42) [inline evap/static/css/_vote-colors.scss]: i'd make that `85%`

File: evap/static/css/_vote-colors.scss [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    position: relative;
+    border-radius: 3px;
+    flex-grow: 1;
+    background: $lighter-gray;
+    box-shadow: inset 0 0 0 1px $light-gray;
+    @include no-user-select;
+    .middle {
+        background: $lighter-gray;
+        position: absolute;
+        $height: 80%;
```
Comment (janno42) [inline evap/static/css/_vote-colors.scss]: i think we can make that `100%`. when the divider doesn't fill the complete vertical space, it adds too much visual noise in my opinion. in full height it fits just perfectly.

File: evap/static/css/_vote-colors.scss [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+        color: $darker-gray;
+        font-size: 0.9rem;
+        @include bar-shadow($white);
+    }
+    .fas {
+
+        color: $white;
+        @include bar-shadow($medium-gray);
+    }
+    &:not(:hover) .adjective, &:hover .fas {
```
Comment (janno42) [inline evap/static/css/_vote-colors.scss]: the hover text needs to be placed a little bit higher. currently it's not correctly centered in chrome (being placed a little bit too much to the bottom is expected in firefox).

**Comment by pixunil** [comment]
I just split the PR in three commits: Refactoring, Implementation and Generation of Test data. I pushed the script for generation into a [gist, for comparison](https://gist.github.com/pixunil/fcdb116dba060e8dc4956c103a87f4c4).

I would love to show that the Refactoring has all tests passed, but I don't know how to nudge Travis.

File: evap/static/css/_vote-colors.scss (hunk 6 above)
Comment (pixunil) [reply evap/static/css/_vote-colors.scss]: Changed `line-height: 1.5rem;` (was `1.6rem`)

File: evap/static/css/_vote-colors.scss (hunk 5 above)
Comment (pixunil) [reply evap/static/css/_vote-colors.scss]: Changed `top: 0; height: 100%`

File: evap/evaluation/migrations/0085_add_bipolar_question_type.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+                    ('T', 'Text Question'),
+                )),
+                ('Unipolar Likert', (
+                    ('L', 'Opinion Question'),
+                )),
+                ('Grade', (
+                    ('G', 'Grade Question'),
+                )),
+                ('Bipolar Likert', (
+                    ('D', 'Easy-Difficult Question'),
```
[the remark this answers is not in the document]
Comment (pixunil) [reply evap/evaluation/migrations/0085_add_bipolar_question_type.py]: Resolved as a `SmallPositiveIntegerField` is now used

File: evap/evaluation/models.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (pixunil) [reply evap/evaluation/models.py]: Also, in the German translation is a non-breaking space. You could change that to a newline for consistency, if you like (but I don't care really)

File: evap/evaluation/migrations/0087_use_integers_and_groups_for_question_type.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
+# Generated by Django 2.0.9 on 2018-10-29 19:48
```
[the remark this answers is not in the document]
Comment (pixunil) [reply evap/evaluation/migrations/0087_use_integers_and_groups_for_question_type.py]: The choices is not a one-dimensional list anymore. It is now a list of groups, where each group has a title and some options. (Groups are Unipolar, Yes-No, Grade, ...)

File: evap/evaluation/migrations/0087_use_integers_and_groups_for_question_type.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+from django.db import migrations, models
+
+
+CONVERSION = {
+    'T': 0,
+    'L': 1,
+    'G': 2,
+    'P': 3,
+    'N': 4,
+    'H': 5
```
[the remark this answers is not in the document]
Comment (pixunil) [reply evap/evaluation/migrations/0087_use_integers_and_groups_for_question_type.py]: Well, I have two constants here: the `str` and `int`. In source would be the `int`, but how do I know if I can import it in the future? Or is this no problem in Django universe?

File: evap/evaluation/models.py [hunk 10]
```diff
@@ -754,7 +770,58 @@ def is_non_grade_rating_question(self):
...
+        _("Disagree"),
+        _("Strongly disagree"),
+        _("no answer")
+    ),
+    Question.GRADE: UnipolarChoices(
+        _("1"),
+        _("2"),
+        _("3"),
+        _("4"),
+        _("5"),
```
[the remark this answers is not in the document]
Comment (pixunil) [reply evap/evaluation/models.py]: I just took the [old one, having the numbers translated](https://github.com/fsr-itse/EvaP/pull/1271/files#diff-33239631e1ff02cda9b3961216dc62c1L23).

File: evap/static/css/evap.scss [hunk 11]
```diff
@@ -625,54 +620,14 @@ td.vote-question.choice-error {
...
 }
-.vote-btn-text {
-    position: relative;
-    top: -1px;
-}
-
-div.vote-inputs {
-    width: 100%;
-    flex-wrap: wrap;
-}
```
[the remark this answers is not in the document]
Comment (pixunil) [reply evap/static/css/evap.scss]: Hey, you found the line that broke the text answers :trophy:. But sadly, it breaks the choiotherces layout as they are too wide. Will work on a working example until push

File: evap/evaluation/migrations/0087_use_integers_and_groups_for_question_type.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (karyon) [reply evap/evaluation/migrations/0087_use_integers_and_groups_for_question_type.py]: good point, they might not exist in the future.

File: evap/evaluation/models.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (janno42) [reply evap/evaluation/models.py]: i'd say we don't need translations for them. i can't think of any situation where we would use a different number depending on the language.

File: evap/static/css/_vote-colors.scss (hunk 3 above)
Comment (pixunil) [reply evap/static/css/_vote-colors.scss]: Changed as the background is now kept gray

**Comment by pixunil** [comment]
As requested, I also partially redesigned the voting layout. Therefore, the css is in some parts very messy (imo). I can not guarantee that every commit does work on its own, but it should be correct and logical

File: evap/evaluation/models.py [hunk 12]
```diff
@@ -718,39 +746,161 @@ def answer_class(self):
...
+
+
+class UnipolarChoices(Choices):
+    values = (1, 2, 3, 4, 5, Choices.NO_ANSWER)
+    colors = ('green', 'lime', 'yellow', 'orange', 'red', 'gray')
+    grades = (1, 2, 3, 4, 5)
+
+
+class BipolarChoices(Choices):
+    cssClass = 'vote-type-bipolar'
```
[the remark this answers is not in the document]
Comment (pixunil) [reply evap/evaluation/models.py]: Update: I use `bipolar` and `unipolar` (later only in comination of both as not-`yes-no`). Only `yes-no` remains unused but is added for completeness

File: evap/evaluation/models.py [hunk 13]
```diff
@@ -771,6 +771,57 @@ def is_heading_question(self):
...
+    Question.GRADE: UnipolarChoices(
+        "1",
+        "2",
+        "3",
+        "4",
+        "5",
+        _("no answer")
+    ),
+    Question.POSITIVE_YES_NO: YesNoChoices(
+        _("Yes"),
```
[the remark this answers is not in the document]
Comment (pixunil) [reply evap/evaluation/models.py]: (capturing this thread for general discussion of values vs grades)

Yep, the classes are just big namedtuples (with `__init__` as some sugar).
But, there are two things that are not mentioned by your solution:
1. Where do you put the css class? Okay, it works in 
2. Where do you put the adjectives for bipolar scales? Like "Easy", "Difficult" without any comparator

For using floats, we could also multiply everything by 3 so we calculate with thirds. Okay, it needs another migration and stuff, but just an idea

[the remark this answers is not in the document]
Comment (karyon) [reply evap/evaluation/models.py]: 1. yes
2. i'd add a BipolarAnswerDataTuple which has the two additional attributes

:)

File: evap/evaluation/models.py [hunk 14]
```diff
@@ -771,6 +771,57 @@ def is_heading_question(self):
         return self.type == self.HEADING
 
 
+class Choices:
+    NO_ANSWER = 6
```
[the remark this answers is not in the document]
Comment (karyon) [reply evap/evaluation/models.py]: yeah okay, perhaps we leave it at that.

File: evap/evaluation/models.py [hunk 15]
```diff
@@ -771,6 +771,57 @@ def is_heading_question(self):
...
+
+    def __init__(self, *names):
+        self.names = names
+
+
+class UnipolarChoices(Choices):
+    values = (1, 2, 3, 4, 5, Choices.NO_ANSWER)
+    colors = ('green', 'lime', 'yellow', 'orange', 'red', 'gray')
```
[the remark this answers is not in the document]
Comment (karyon) [reply evap/evaluation/models.py]: alright... i think i would put the full names here, no need for an additional mapping, or is there?

File: evap/evaluation/templatetags/evaluation_filters.py [hunk 16]
```diff
@@ -57,13 +57,10 @@ def percentage_value(fraction, population):
...
 @register.filter
-def get_answer_name(question, grade):
-    choices = Choices.QUESTION_TYPES[question.type]
-    try:
-        idx = choices.grades.index(grade)
-        return choices.names[idx]
-    except (IndexError, ValueError):
-        return None
+def to_colors(choices):
+    if choices:
```
[the remark this answers is not in the document]
Comment (karyon) [reply evap/evaluation/templatetags/evaluation_filters.py]: hm, not sure. perhaps just add a comment that explains in what case the fallback is used?

File: evap/results/tools.py [hunk 17]
```diff
@@ -158,6 +164,19 @@ def normalized_distribution(distribution):
...
 
 
+def unipolarized_distribution(result):
+    summed_distribution = [0, 0, 0, 0, 0]
+
+    for value, grade in zip(result.counts, result.choices.grades):
+        grade_fraction, grade = modf(grade)
+        grade = int(grade)
+        summed_distribution[grade - 1] += (1 - grade_fraction) * value
+        if grade < 5:
```
[the remark this answers is not in the document]
Comment (karyon) [reply evap/results/tools.py]: I really don't know what's better :)  i'll just resolve this, it not that important.

File: evap/evaluation/models.py [hunk 18]
```diff
@@ -687,13 +687,41 @@ def is_general(self):
...
-        ("L", _("Likert Question")),
-        ("G", _("Grade Question")),
-        ("P", _("Positive Yes-No Question")),
-        ("N", _("Negative Yes-No Question")),
-        ("H", _("Heading")),
+        (_("Text"), (
+            (TEXT, _("Text Question")),
+        )),
+        (_("Unipolar Likert"), (
+            (LIKERT, _("Opinion Question")),
```
Comment (janno42) [inline evap/evaluation/models.py]: `Agreement question` (and the other ones also with a lowercase `q`) please

File: evap/evaluation/migrations/0090_add_bipolar_question_type.py [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+                        (1, 'Opinion Question'),
+                    )),
+                    ('Grade', (
+                        (2, 'Grade Question'),
+                    )),
+                    ('Bipolar Likert', (
+                        (6, 'Easy-Difficult Question'),
+                        (7, 'Few-Many Question'),
+                        (8, 'Little-Much Question'),
+                        (9, 'Small-Big Question'),
```
Comment (janno42) [inline evap/evaluation/migrations/0090_add_bipolar_question_type.py]: `Small-large question`

File: evap/static/css/_vote-elements.scss [hunk 20]
```diff
@@ -0,0 +1,216 @@
...
+    .vote-answer-buttons {
+        padding-top: 0.5rem;
+    }
+    .vote-inputs {
+        justify-content: flex-end;
+    }
+    .vote-btn {
+        margin-top: 0;
+    }
+    .vote-type-unipolar > .vote-btn:not(:last-child), .vote-type-bipolar > .vote-btn:not(:last-child) {
```
Comment (janno42) [inline evap/static/css/_vote-elements.scss]: in two lines please

**Comment by karyon** [comment]
regarding ditching the values tuples: we determined that the browser shouldn't make any problems because it gets a string and should send the same string back. an open question was whether `some_float == float(str(some_float))` would always work in python. [the answer is yes](https://stackoverflow.com/questions/25898733/why-does-strfloat-return-more-digits-in-python-3-than-python-2). to be extra safe and for more explicitness, we could use `some_float == float(repr(some_float))`.

as to where to put the NO_ANSWER, we could add that directly in the form along with the `_("no answer")`. While that would remove the option to create questions without "no answer" choice, right now we don't need that option.

while i still have the feeling it would be the right thing, i also can't get rid of the feeling that it's not a really good idea at the same time. if i'm the only one actually considering trying it, we should maybe not do it :)

**Comment by janno42** [comment]
i also have mixed feelings about the floats and while i see the advantage of dropping one value i'm not quite sure it's worth having a kind of strange identifier (which is only strange because we're not used to it, but still). so if i have to cast a vote i would say let's not use the float identifiers.

**Comment by He3lixxx** [comment]
About the floats: What happens if Python decides to switch from IEEE754/64bit to a different representation? Is the text-representation stored somewhere permanently? If it is not, I'm perfectly fine with the floats because I don't think it will ever break. People have used UUIDs as post parameters, that also works.

File: evap/static/css/_vote-elements.scss [hunk 21]
```diff
@@ -0,0 +1,222 @@
...
+    green: $evap-green,
+    lime: $evap-lime,
+    yellow: $evap-yellow,
+    orange: $evap-orange,
+    red: $evap-red,
+    gray: $medium-gray
+);
+
+.vote-questionnaire .row:nth-child(odd) {
+    background-color: #f2f2f2; // FIXME
```
[the remark this answers is not in the document]
Comment (pixunil) [reply evap/static/css/_vote-elements.scss]: Relict :trophy:

File: evap/evaluation/models.py [hunk 22]
```diff
@@ -776,6 +777,63 @@ def is_heading_question(self):
...
+    'grades': (1, 2, 3, 4, 5)
+}
+
+BASE_YES_NO_CHOICES = {
+    'values': (1, 5, Choices.NO_ANSWER),
+    'colors': ('green', 'red', 'gray'),
+    'grades': (1, 5)
+}
+
+Choices.QUESTION_TYPES = {
```
[the remark this answers is not in the document]
Comment (pixunil) [reply evap/evaluation/models.py]: Okay, at least it's not global like C does, but module global. I'm also fine with it

File: evap/evaluation/templatetags/evaluation_filters.py [hunk 23]
```diff
@@ -52,15 +57,10 @@ def percentage_value(fraction, population):
...
-    elif question.is_positive_yes_no_question:
-        return POSITIVE_YES_NO_NAMES.get(grade)
-    elif question.is_negative_yes_no_question:
-        return NEGATIVE_YES_NO_NAMES.get(grade)
-    else:
-        return grade
+def to_colors(choices):
+    if choices:
+        return choices.colors
+    return BASE_UNIPOLAR_CHOICES['colors']
```
[the remark this answers is not in the document]
Comment (pixunil) [reply evap/evaluation/templatetags/evaluation_filters.py]: I rotated the branches, it jsut looked more logical to me

**Comment by janno42** [comment]
Thank you again for this amazing PR! You invested a lot of time and I'd like to give you kudos for that, I hope you also enjoyed working on it. I very much appreciate your focus on details, the nice UI elements we now have and the new possibilities this PR offers for our usage of EvaP in 2019 and the coming years. Thanks! 🎆

**Comment by karyon** [comment]
10/10 would review again

**Pull request merged**